test/unittest/memory/unittest_cache_loader.cpp

```diff
@@ -0,0 +1,417 @@
+  EXPECT_NO_THROW(mem3 = pool->getMemory(idx3));
+  EXPECT_NE(mem1, nullptr);
+  EXPECT_EQ(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2, nullptr);
+  EXPECT_EQ(mem2->getAddr(), nullptr);
+  EXPECT_NE(mem3, nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+
+  // Check memory loading for each execution order
+  loader->load(1);
+  EXPECT_NE(mem1->getAddr(), nullptr);
+  EXPECT_EQ(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(2);
+  EXPECT_NE(mem1->getAddr(), nullptr);
+  EXPECT_EQ(mem2->getAddr(), nullptr);
+  EXPECT_NE(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(3);
+  EXPECT_NE(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_NE(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(4);
+  EXPECT_NE(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_NE(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(5);
+  EXPECT_NE(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(6);
+  EXPECT_EQ(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(7);
+  EXPECT_EQ(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(8);
+  EXPECT_EQ(mem1->getAddr(), nullptr);
+  EXPECT_NE(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+
+  loader->load(9);
+  EXPECT_EQ(mem1->getAddr(), nullptr);
+  EXPECT_EQ(mem2->getAddr(), nullptr);
+  EXPECT_EQ(mem3->getAddr(), nullptr);
+  pool->flush();
+}
+
+/**
+ * @brief load asynchronously
+ */
+TEST_F(CacheLoaderTest, load_async_01_p) {
+  std::shared_ptr<nntrainer::MemoryData<float>> mem;
+  auto idx = pool->requestMemory(4, 1, 5, {1, 2, 3, 4, 5});
+  EXPECT_NO_THROW(pool->planLayout(nntrainer::OptimizedV1Planner()));
+  EXPECT_NO_THROW(pool->allocate());
+  EXPECT_EQ(pool->size(), 4);
+  EXPECT_NO_THROW(mem = pool->getMemory(idx));
+  EXPECT_NE(mem, nullptr);
+  EXPECT_EQ(mem->getAddr(), nullptr);
+
+  std::promise<void> p;
+  auto future = p.get_future();
+  auto complete = [&p, &mem](int id,
+                             nntrainer::TaskExecutor::CompleteStatus status) {
+    EXPECT_EQ(status, nntrainer::TaskExecutor::SUCCESS);
+    EXPECT_NE(mem->getAddr(), nullptr);
+    p.set_value();
+  };
+
+  // Check memory loading for each execution order
+  int id = loader->loadAsync(1, complete);
+  EXPECT_EQ(mem->getAddr(), nullptr);
+
+  future.wait();
+  pool->flush();
+}
+
+/**
+ * @brief load asynchronously
+ */
+TEST_F(CacheLoaderTest, load_async_02_p) {
+  std::shared_ptr<nntrainer::MemoryData<float>> mem1, mem2, mem3;
+  auto idx1 = pool->requestMemory(4, 1, 5, {1, 2, 3, 4, 5});
+  auto idx2 = pool->requestMemory(4, 3, 8, {3, 4, 5, 6, 7, 8});
```

Comment:
I think it might be good if you could add about non-continuous execution order lists later.

Reply:
thanks. `load_async_03_p` case is added.